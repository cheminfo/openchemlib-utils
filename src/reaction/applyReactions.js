import { appendOCLReaction } from './utils/appendOCLReaction.js';
import { applyOneReactantReactions } from './utils/applyOneReactantReactions.js';
import { groupTreesByProducts } from './utils/groupTreesByProducts.js';

/**
 * @typedef {object} ReactionEntry
 * @property {string} label
 * @property {string} rxnCode
 */

/**
 * Create reaction trees of products based on reactions and reactants
 * @param {import('openchemlib').Molecule[]} reactants
 * @param {ReactionEntry[]} reactions - array of reactions objects with rxnCode, label
 * @param {object} options - options to apply the reaction
 * @param {number} [options.maxDepth=5] - max depth of the recursion
 * @param {number} [options.limitReactions=200] - limit the number of reactions to apply
 * @param {boolean} [options.getProductsTrees=false] - if true, the returned object will have a products property with the products trees grouped by idCode else it will be an empty array
 * @returns {object} The returned object has two properties:
 * - trees: the tree of reactions
 * - products: reactions trees grouped by product idCode
 */
export function applyReactions(reactants, reactions, options = {}) {
  // eslint-disable-next-line no-console
  console.warn('applyReactions is deprecated');
  // Reaction are applied recursively until maximal tree depth is reached (default 10)
  const {
    maxDepth = 5,
    limitReactions = 200,
    getProductsTrees = false,
  } = options;
  const moleculesInfo = new Map();
  const processedMolecules = new Set();
  if (reactants.length === 0) {
    throw new Error('Can not extract OCL because there is no reactants');
  }
  // get the OCL object from the first reactant
  const OCL = reactants[0].getOCL();

  reactions = appendOCLReaction(reactions, OCL);

  const stats = { counter: 0 };
  const trees = [];
  // Start the recursion by applying the first level of reactions
  let todoCurrentLevel = applyOneReactantReactions(reactants, reactions, {
    OCL,
    currentDepth: 0,
    moleculesInfo,
    processedMolecules,
    maxDepth,
    trees,
    stats,
    limitReactions,
  });
  do {
    const nexts = [];
    for (const todo of todoCurrentLevel) {
      nexts.push(todo());
    }
    todoCurrentLevel = nexts.flat();
  } while (todoCurrentLevel.length > 0);
  let products;
  if (getProductsTrees) {
    products = groupTreesByProducts(trees);
  } else {
    products = [];
  }
  return { trees, products, stats };
}
