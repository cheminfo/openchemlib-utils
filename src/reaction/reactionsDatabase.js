export const reactionsDatabase=[{"rxnCode":"gGQ@@eKtrA@!gGQ@@djmRA@#qtJk qtSi#!Ruwu~@M_|upKWoQwr@` !RuwwW?Gx@bOrH_QuZ@`","Label":"Claisen rearrangement"},{"rxnCode":"gGP@DkUjPH!gGP@DkUjPH#qbAk q`Si#!Ru?sH@gx@upKWoQvJP` !Rr@KW?Gx@bOrH_QwZP`","Label":"Cope rearrangement"},{"rxnCode":"gC`@Die@` eF@HhP!gFp@DiTujhB#qMs Qd qi^T#!R@Fq?[@@SGRk| !R@AL]lkp !R?`Bw?[\\BmpJw?Qtz@`","Label":"Diels-Alder"},{"rxnCode":"eMBBHRZB eFB`HNcA@!gC`H@DISTA@#QF qN qJt#!R@Fp@Dqur_@ !R_vp]rgp !R_vp@[G|SGSi|","Label":"SN2 thioether"}]