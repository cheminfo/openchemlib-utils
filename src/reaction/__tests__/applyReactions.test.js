import { Molecule } from 'openchemlib/full.js';

import { applyReactions } from '../applyReactions.js';

import { reactionsDatabase } from './reactionsDatabase.js';

describe('applyReactions', () => {
  it('ethanol', () => {
    const ethanol = Molecule.fromSmiles('CCO');
    const results = applyReactions([ethanol], reactionsDatabase, {});
    expect(results).toHaveLength(1);
    const firstResult = results[0];
    expect(firstResult.products).toHaveLength(1);
    const firstProduct = firstResult.products[0];
    expect(firstProduct.mf).toBe('C2H6S');
    expect(firstProduct.children).toHaveLength(1);
    const firstChild = firstProduct.children[0];
    expect(firstChild.reactant.mf).toBe('C2H6S');
    expect(firstChild.products).toHaveLength(1);
    expect(firstChild.products[0].mf).toBe('C3H8S');
  });

  it('ethylene glycol', () => {
    const diol = Molecule.fromSmiles('OCCO');
    const results = applyReactions([diol], reactionsDatabase, {});
    expect(results).toHaveLength(1);
    const firstResult = results[0];
    expect(firstResult.products).toHaveLength(1);
    const firstProduct = firstResult.products[0];
    expect(firstProduct.mf).toBe('C2H6OS');
    expect(firstProduct.children).toHaveLength(2);

    const firstChild = firstProduct.children[0];
    expect(firstChild.reactant.mf).toBe('C2H6OS');
    expect(firstChild.products).toHaveLength(1);
    expect(firstChild.products[0].mf).toBe('C2H6S2');
    const secondChild = firstProduct.children[1];
    expect(secondChild.reactant.mf).toBe('C2H6OS');
    expect(secondChild.products).toHaveLength(1);
    expect(secondChild.products[0].mf).toBe('C3H8OS');

    expect(results).toMatchSnapshot()
  });

  it('propane-1,2-diol', () => {
    const propanediol = Molecule.fromSmiles('CC(O)CO');
    const results = applyReactions([propanediol], reactionsDatabase, {});
    expect(results).toHaveLength(2);
    const firstResult = results[0];
    expect(firstResult.products).toHaveLength(1);
    const firstProduct = firstResult.products[0];
    expect(firstProduct.mf).toBe('C3H8OS');
    expect(firstProduct.children).toHaveLength(2);
    const secondResult = results[1];
    expect(secondResult.products).toHaveLength(1);
    const secondProduct = secondResult.products[0];
    expect(secondProduct.mf).toBe('C3H8OS');
    expect(secondProduct.children).toHaveLength(2);

    const firstChild = firstProduct.children[0];
    expect(firstChild.reactant.mf).toBe('C3H8OS');
    expect(firstChild.products).toHaveLength(1);
    expect(firstChild.products[0].mf).toBe('C3H8S2');
    const secondChild = firstProduct.children[1];
    expect(secondChild.reactant.mf).toBe('C3H8OS');
    expect(secondChild.products).toHaveLength(1);
    expect(secondChild.products[0].mf).toBe('C4H10OS');
    expect(secondChild.reaction.oclReaction).toBeUndefined();
    expect(secondChild.products[0].children[0].products[0].mf).toBe('C4H10S2');
  });

  it('propane-1,2-diol maxDepth: 1', () => {
    const propanediol = Molecule.fromSmiles('CC(O)CO');
    const results = applyReactions([propanediol], reactionsDatabase, {
      maxDepth: 1,
    });
    expect(results).toHaveLength(2);
    const firstResult = results[0];
    expect(firstResult.products).toHaveLength(1);
    const firstProduct = firstResult.products[0];
    expect(firstProduct.mf).toBe('C3H8OS');
    expect(firstProduct.children).toHaveLength(0);
  });

  it('propane-1,2-diol maxDepth: 2', () => {
    const propanediol = Molecule.fromSmiles('CC(O)CO');
    const results = applyReactions([propanediol], reactionsDatabase, {
      maxDepth: 2,
    });
    expect(results).toHaveLength(2);
    const firstResult = results[0];
    expect(firstResult.products).toHaveLength(1);
    const firstProduct = firstResult.products[0];
    expect(firstProduct.mf).toBe('C3H8OS');
    const mfs = firstProduct.children.map((child) => child.products[0].mf);
    expect(mfs).toStrictEqual(['C3H8S2', 'C4H10OS']);
  });
});
