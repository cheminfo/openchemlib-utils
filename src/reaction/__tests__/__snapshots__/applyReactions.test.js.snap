// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyReactions ethylene glycol 1`] = `
[
  {
    "products": [
      {
        "children": [
          {
            "products": [
              {
                "children": [
                  {
                    "products": [
                      {
                        "children": [
                          {
                            "products": [
                              {
                                "children": [],
                                "idCode": "gGPD@DDHRgZj@@",
                                "mf": "C4H10S2",
                                "molfile": "
Actelion Java MolfileCreator 1.0

  6  5  0  0  0  0  0  0  0  0999 V2000
   -0.0191   -0.1275   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.1530   -1.8268   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.1315    0.2168   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    4.5318   -1.6197   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    0.1517   -0.0036   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.6925   -1.8283   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  1  1  0  0  0  0
  3  5  1  0  0  0  0
  4  2  1  0  0  0  0
  4  6  1  0  0  0  0
M  END
",
                              },
                            ],
                            "reactant": {
                              "idCode": "gJPD@DDHRgj\`@",
                              "mf": "C3H8S2",
                              "molfile": "
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
    3.2986   -1.4727   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0207   -0.0078   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.6674   -1.2710   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    0.1205    0.2225   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    3.7988   -1.4741   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  4  2  1  0  0  0  0
  3  1  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
",
                            },
                            "reaction": {
                              "Label": "SH to SMe",
                              "rxnCode": "eFB@HcAa@!eMB@HchH#S\\ S\\@#!R@FL]bgp !R@FL?XqwZ_@",
                            },
                          },
                        ],
                        "idCode": "gJPD@DDHRgj\`@",
                        "mf": "C3H8S2",
                        "molfile": "
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
    3.2986   -1.4727   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0207   -0.0078   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.6674   -1.2710   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    0.1205    0.2225   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    3.7988   -1.4741   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  4  2  1  0  0  0  0
  3  1  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
",
                      },
                    ],
                    "reactant": {
                      "idCode": "gC\`D@DXHRVh@",
                      "mf": "C2H6S2",
                      "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
    0.1156   -0.0433   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.7862   -1.0435   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.2077    0.1916   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    3.0661   -0.8861   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  1  1  0  0  0  0
  4  2  1  0  0  0  0
M  END
",
                    },
                    "reaction": {
                      "Label": "SH to SMe",
                      "rxnCode": "eFB@HcAa@!eMB@HchH#S\\ S\\@#!R@FL]bgp !R@FL?XqwZ_@",
                    },
                  },
                ],
                "idCode": "gC\`D@DXHRVh@",
                "mf": "C2H6S2",
                "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
    0.1156   -0.0433   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.7862   -1.0435   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.2077    0.1916   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    3.0661   -0.8861   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  1  1  0  0  0  0
  4  2  1  0  0  0  0
M  END
",
              },
            ],
            "reactant": {
              "idCode": "gCaHL@aIZ\`@",
              "mf": "C2H6OS",
              "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
   19.5729   -7.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   19.5617   -8.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   20.4220   -8.8405   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   20.4444   -6.8406   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  4  1  1  0  0  0  0
M  END
",
            },
            "reaction": {
              "Label": "OH by SH",
              "rxnCode": "eFHBLGtV!eFB@HcA}E\`#aP aP#!R@AL]\\mp !R@AL]Nmp",
            },
          },
          {
            "products": [
              {
                "children": [
                  {
                    "products": [
                      {
                        "children": [],
                        "idCode": "gJPD@DDHRgj\`@",
                        "mf": "C3H8S2",
                        "molfile": "
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
    0.1021   -0.1776   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.7040   -1.3873   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.2407    0.1724   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    0.3117   -0.0073   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.0044   -1.2183   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  1  1  0  0  0  0
  3  4  1  0  0  0  0
  5  2  1  0  0  0  0
M  END
",
                      },
                    ],
                    "reactant": {
                      "idCode": "gJQHB@aJ^j@@",
                      "mf": "C3H8OS",
                      "molfile": "
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
   18.7877   -8.0008   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   18.7645   -9.0005   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   19.6187   -9.5204   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   19.6757   -7.5152   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
   20.5359   -8.0008   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  4  1  1  0  0  0  0
  4  5  1  0  0  0  0
M  END
",
                    },
                    "reaction": {
                      "Label": "OH by SH",
                      "rxnCode": "eFHBLGtV!eFB@HcA}E\`#aP aP#!R@AL]\\mp !R@AL]Nmp",
                    },
                  },
                ],
                "idCode": "gJQHB@aJ^j@@",
                "mf": "C3H8OS",
                "molfile": "
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
   18.7877   -8.0008   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   18.7645   -9.0005   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   19.6187   -9.5204   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   19.6757   -7.5152   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
   20.5359   -8.0008   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  4  1  1  0  0  0  0
  4  5  1  0  0  0  0
M  END
",
              },
            ],
            "reactant": {
              "idCode": "gCaHL@aIZ\`@",
              "mf": "C2H6OS",
              "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
   19.5729   -7.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   19.5617   -8.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   20.4220   -8.8405   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   20.4444   -6.8406   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  4  1  1  0  0  0  0
M  END
",
            },
            "reaction": {
              "Label": "SH to SMe",
              "rxnCode": "eFB@HcAa@!eMB@HchH#S\\ S\\@#!R@FL]bgp !R@FL?XqwZ_@",
            },
          },
        ],
        "idCode": "gCaHL@aIZ\`@",
        "mf": "C2H6OS",
        "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
   19.5729   -7.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   19.5617   -8.3308   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   20.4220   -8.8405   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   20.4444   -6.8406   -0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  4  1  1  0  0  0  0
M  END
",
      },
    ],
    "reactant": {
      "idCode": "gC\`\`Adej@@",
      "mf": "C2H6O2",
      "molfile": "
Actelion Java MolfileCreator 1.0

  4  3  0  0  0  0  0  0  0  0999 V2000
    2.5981   -0.0000   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660   -0.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.5000   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
M  END
",
    },
    "reaction": {
      "Label": "OH by SH",
      "rxnCode": "eFHBLGtV!eFB@HcA}E\`#aP aP#!R@AL]\\mp !R@AL]Nmp",
    },
  },
]
`;
